src/promptflow/promptflow/_utils/openai_metrics_calculator.py

```diff
@@ -0,0 +1,159 @@
+import tiktoken
+
+
+class OpenAIMetricsCalculator:
+    def __init__(self, logger=None) -> None:
+        self._logger = logger
+
+    def get_openai_metrics_from_api_call(self, api_call: dict):
+        total_metrics = {}
+        if self._need_collect_metrics(api_call):
+            metrics = self._get_openai_metrics_for_signal_api(api_call)
+            self.merge_metrics_dict(total_metrics, metrics)
+
+        children = api_call.get("children")
+        if children is not None:
+            for child in children:
+                child_metrics = self.get_openai_metrics_from_api_call(child)
+                self.merge_metrics_dict(total_metrics, child_metrics)
+
+        return total_metrics
+
+    def _need_collect_metrics(self, api_call: dict):
+        if api_call.get("type") != "LLM":
+            return False
+        output = api_call.get("output")
+        if not isinstance(output, dict) and not isinstance(output, list):
+            return False
+        return True
+
+    def _get_openai_metrics_for_signal_api(self, api_call: dict):
+        output = api_call.get("output")
+        if isinstance(output, dict):
+            usage = output.get("usage")
+            if isinstance(usage, dict):
+                return usage
+            self._log_warning("Cannot find usage in output, will calculate metrics from response data directly.")
+
+        name = api_call.get("name")
+        if name.split(".")[-2] == "ChatCompletion":
+            return self._get_openai_metrics_for_chat_api(api_call)
+        elif name.split(".")[-2] == "Completion":
+            return self._get_openai_metrics_for_completion_api(api_call)
+        else:
+            self._log_warning(f"Cannot calculate metrics for api: {name}.")
+            return {}
+
+    def _get_openai_metrics_for_chat_api(self, api_call):
+        inputs = api_call.get("inputs")
+        output = api_call.get("output")
+        metrics = {}
+        enc, tokens_per_message, tokens_per_name = self._get_encoding_for_chat_api(inputs["engine"])
```

Comment:
We should use api_type to decide whether use engine or model

Reply:
Updated.